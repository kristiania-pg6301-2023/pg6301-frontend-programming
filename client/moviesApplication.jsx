import React, { useEffect, useState } from "react";
import { Route, Routes } from "react-router-dom";

export function ListMovies({ movies }) {
  return (
    <>
      <h2>All movies</h2>
      {movies.map((m) => (
        <div key={m.title}>
          <h3>{m.title}</h3>
        </div>
      ))}
    </>
  );
}

function AddMovie() {
  return (
    <form method="dialog">
      <h2>Movies</h2>
      <div>
        Title: <br />
        <input />
      </div>
      <div>
        <button>Submit</button>
      </div>
    </form>
  );
}

export function MoviesRoutes({ fetchMovies }) {
  const [movies, setMovies] = useState([]);

  async function loadMovies() {
    const movies = await fetchMovies();
    setMovies(movies);
  }

  useEffect(() => {
    loadMovies();
  }, []);

  return (
    <Routes>
      <Route path={"/"} element={<ListMovies movies={movies} />} />
      <Route path={"/movies/new"} element={<AddMovie />} />
      <Route path={"*"} element={<h2>Not Found</h2>} />
    </Routes>
  );
}
