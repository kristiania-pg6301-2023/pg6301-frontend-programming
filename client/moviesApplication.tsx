import React, { Dispatch, SetStateAction, useEffect, useState } from "react";
import { Route, Routes, useNavigate } from "react-router-dom";

interface Movie {
  _id: string;
  title: string;
  plot: string;
  year: number;
}

interface MovieParameters {
  countries: string[];
  years: number[];
}

export interface MovieQuery {
  countries?: string;
  year?: number;
}

export function ListMovies({
  movies,
  parameters,
  setQuery,
}: {
  movies: Movie[];
  parameters: MovieParameters;
  query: MovieQuery;
  setQuery: Dispatch<SetStateAction<MovieQuery>>;
}) {
  return (
    <>
      <h2>All movies</h2>
      <div className={"query-filter"}>
        <div>
          <label>
            Country: <br />
            <select
              onChange={(e) =>
                setQuery((old) => ({ ...old, countries: e.target.value }))
              }
            >
              <option></option>
              {parameters.countries.map((c) => (
                <option key={c}>{c}</option>
              ))}
            </select>
          </label>
        </div>
        <div>
          <label>
            Year: <br />
            <select
              onChange={(e) =>
                setQuery((old) => ({ ...old, year: parseInt(e.target.value) }))
              }
            >
              <option></option>
              {parameters.years.map((y) => (
                <option key={y}>{y}</option>
              ))}
            </select>
          </label>
        </div>
      </div>
      {movies.map((m) => (
        <div key={m._id}>
          <h3>{m.title}</h3>
          <p>{m.plot}</p>
        </div>
      ))}
    </>
  );
}

function AddMovie({ onCreate }: { onCreate(movie: Omit<Movie, "_id">): void }) {
  const [title, setTitle] = useState("");
  const [year, setYear] = useState("");
  const [plot, setPlot] = useState("");

  async function handleSubmit() {
    await onCreate({ title, plot, year: parseInt(year) });
  }

  return (
    <form method="dialog" onSubmit={handleSubmit}>
      <h2>Movies</h2>
      <div>
        Title: <br />
        <input value={title} onChange={(e) => setTitle(e.target.value)} />
      </div>
      <div>
        Year: <br />
        <input value={year} onChange={(e) => setYear(e.target.value)} />
      </div>
      <div>
        Plot: <br />
        <textarea value={plot} onChange={(e) => setPlot(e.target.value)} />
      </div>
      <div>
        <button>Submit</button>
      </div>
    </form>
  );
}

export interface MoviesRoutesProps {
  fetchMovies(query: MovieQuery): Promise<Movie[]>;

  fetchParameters(): Promise<MovieParameters>;

  insertMovie(movie: Movie): Promise<void>;
}

export function MoviesRoutes({
  fetchMovies,
  fetchParameters,
  insertMovie,
}: MoviesRoutesProps) {
  const navigate = useNavigate();
  const [movies, setMovies] = useState<Movie[]>([]);
  const [query, setQuery] = useState<MovieQuery>({});
  const [parameters, setParameters] = useState<MovieParameters>({
    countries: [],
    years: [],
  });

  async function loadMovies() {
    setMovies(await fetchMovies(query));
    setParameters(await fetchParameters());
  }

  useEffect(() => {
    loadMovies();
  }, [query]);

  async function handleCreate(movie: Movie) {
    await insertMovie(movie);
    await loadMovies();
    navigate("/");
  }

  return (
    <Routes>
      <Route
        path={"/"}
        element={
          <ListMovies
            movies={movies}
            parameters={parameters}
            query={query}
            setQuery={setQuery}
          />
        }
      />
      <Route
        path={"/movies/new"}
        element={<AddMovie onCreate={handleCreate} />}
      />
      <Route path={"*"} element={<h2>Not Found</h2>} />
    </Routes>
  );
}
