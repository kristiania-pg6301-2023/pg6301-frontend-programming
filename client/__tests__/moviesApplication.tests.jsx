import renderer, { act } from "react-test-renderer";
import { MoviesRoutes } from "../moviesApplication";
import { MemoryRouter } from "react-router-dom";

const movies = [
  { title: "Barbie", year: "2023", plot: "Adventures in Pink" },
  { title: "Oppenheimer", year: "2023", plot: "The Manhattan project" },
];

describe("movies database view", () => {
  it("matches snapshot", async () => {
    let component;
    await act(async () => {
      component = renderer.create(
        <MemoryRouter>
          <MoviesRoutes fetchMovies={() => movies} />,
        </MemoryRouter>,
      );
    });
    expect(component).toMatchSnapshot();
    expect(
      component.root.findAllByType("h3").map((c) => c.children.join(" ")),
    ).toEqual(["Barbie", "Oppenheimer"]);
  });

  it("shows new movies form", async () => {
    let component;
    await act(async () => {
      component = renderer.create(
        <MemoryRouter initialEntries={["/movies/new"]}>
          <MoviesRoutes fetchMovies={jest.fn()} />,
        </MemoryRouter>,
      );
    });
    expect(component).toMatchSnapshot();
    expect(component.root.findByType("button").children.join(" ")).toEqual(
      "Submit",
    );
  });
});
