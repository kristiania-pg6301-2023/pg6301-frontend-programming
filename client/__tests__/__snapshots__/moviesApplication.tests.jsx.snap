// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`movies database view matches snapshot 1`] = `
[
  <h2>
    All movies
  </h2>,
  <div>
    <h3>
      Barbie
    </h3>
  </div>,
  <div>
    <h3>
      Oppenheimer
    </h3>
  </div>,
  ",",
]
`;

exports[`movies database view shows new movies form 1`] = `
[
  <form
    method="dialog"
  >
    <h2>
      Movies
    </h2>
    <div>
      Title: 
      <br />
      <input />
    </div>
    <div>
      <button>
        Submit
      </button>
    </div>
  </form>,
  ",",
]
`;
